api/deploy.go

```diff
@@ -86,44 +86,58 @@ func newServiceDeployer(api *API, env map[string]string, options ...DeployServic
 
 // FromURL deploys a service from git or tarball url.
 func (d *serviceDeployer) FromURL(url string) (*service.Service, *importer.ValidationError, error) {
+	path, err := d.createTempDir()
+	if err != nil {
+		return nil, nil, err
+	}
+	d.sendStatus("Downloading service...", Running)
 	defer d.closeStatus()
 	if xgit.IsGitURL(url) {
-		d.sendStatus("Downloading service...", Running)
-		path, err := d.createTempDir()
-		if err != nil {
-			return nil, nil, err
-		}
 		defer os.RemoveAll(path)
 		if err := xgit.Clone(url, path); err != nil {
 			return nil, nil, err
 		}
 
-		// XXX: remove .git folder from repo.
-		// It makes docker build iamge id same between repo clones.
-		if err := os.RemoveAll(filepath.Join(path, ".git")); err != nil {
+	} else {
+		// if not git repo then it should be tarball
+		resp, err := http.Get(url)
+		if err != nil {
 			return nil, nil, err
 		}
-
-		d.sendStatus("Service downloaded with success", DonePositive)
-		r, err := xarchive.GzippedTar(path, nil)
-		if err != nil {
+		defer resp.Body.Close()
+		if err := archive.Untar(resp.Body, path, nil); err != nil {
 			return nil, nil, err
 		}
-		defer r.Close()
-		return d.deploy(r)
 	}
 
-	// if not git repo then it should be tarball
-	resp, err := http.Get(url)
+	// XXX: remove .git folder from repo.
+	// It makes docker build iamge id same between repo clones.
+	if err := os.RemoveAll(filepath.Join(path, ".git")); err != nil {
+		return nil, nil, err
+	}
+
+	// NOTE: this is check for tar repos, if there is only one
+	// directory inside untar archive set temp path to it.
+	dirs, err := ioutil.ReadDir(path)
 	if err != nil {
 		return nil, nil, err
 	}
-	defer resp.Body.Close()
-	return d.deploy(resp.Body)
+	if len(dirs) == 1 && dirs[0].IsDir() {
+		path = filepath.Join(path, dirs[0].Name())
+	}
+
+	d.sendStatus("Service downloaded with success", DonePositive)
+	r, err := archive.Tar(path, archive.Uncompressed)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer r.Close()
+	return d.deploy(r)
 }
 
 // FromArchive deploys a service from a gzipped tarball.
 func (d *serviceDeployer) FromArchive(r io.Reader) (*service.Service, *importer.ValidationError, error) {
+	d.sendStatus("Receiving service context...", Running)
```

Comment:
I think it's better to move this _service_ pkg and also keep the _Service context received (done)_ message. This way we can show the status info about sending service context for both _FromArchive()_ & _FromURL()_

Reply:
the service should not be aware of the service deployer. I'm totally fine to keep it that way.